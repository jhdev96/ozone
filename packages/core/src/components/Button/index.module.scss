@import '../../scss/root';
@import '../../scss/mixins/button/button.scss';
@import '../../scss/mixins/button/variables.scss';
                                  

// For styling `flat` buttons
$flatMap: (
  primary: (
    color: #fff,
    bg-color: var(--blue-3)
  ),
  secondary: (
    color: #fff,
    bg-color: var(--red-3)
  ),
  warning: (
    color: var(--black-core),
    bg-color: var(--yellow-3)
  )
);

/* Major button classes - primary, secondary, warning, basic */                                   
.primary {
  @include button;

  color: #fff;
  background-color: var(--blue-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;

    color: var(--blue-core);
    border: 1px solid var(--blue-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--red-core) auto 1px !important;
  }
}

.secondary {
  @include button;

  color: #fff;
  background-image: $secondary-gradient;
  background-repeat: repeat-x;
  border-color: $button-bc1 $button-bc2 $button-bc3 !important;
  box-shadow: inset 0 1px 0 $primary-shade1, 
                    0 1px 2px $primary-shade2;
  background-color: var(--red-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;
    color: var(--red-core);
    border: 1px solid var(--red-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--blue-core) auto 1px !important;
  }
}

.warning {
  @include button;

  color: var(--black-core);
  background-color: var(--yellow-core);
  border-radius: 4px;

  &:focus {
    outline: var(--black-core) auto 1px !important;
  }
}

.basic {
  @include button;

  border-radius: 3px;

  &:focus {
    outline: var(--gray-4) auto 1px !important;
  }
}

/* The following styles serve as modifiers 
 * for the major button classes 
*/
.flat {
  background-image: none !important;
  background-repeat: no-repeat;
  border: 0;
  border-radius: 0 !important;
  box-shadow: none !important;
  text-shadow: none !important;

  @each $class, $rule in $flatMap {
    &.#{$class}:not(.disabled):not(.loading) {
      &:hover {
        color: map-get($rule, 'color');
        background-color: map-get($rule, 'bg-color');
      }
    }
  }
}

.fill { width: 100%; }

.loading {
  padding: 8px 20px 5px 20px;
  max-height: 35px;
  opacity: 0.90;
  cursor: default;

  &:focus {outline: none !important;}
}

.disabled {
  opacity: 0.70;
  cursor: default;
}

.small {
  min-width: fit-content;
  padding: 8px !important;
}

.medium { min-width: 40px; }

.large {
  min-width: 70px;
  max-height: 70px !important;
  padding: 15px 20px !important;
  font-size: 1.2em !important;
}

.light {
  &.basic {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }
}

.dark {
  &.basic {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}