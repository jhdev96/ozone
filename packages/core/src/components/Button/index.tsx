import React from 'react';
import classNames from '../../utils/classNames';
import { 
  ButtonVariation,
  loaderTheme
} from './utils';
import LoadingDots from '../LoadingDots';
import styles from './index.module.scss';


export interface ButtonProps {
  /** 
   * The main classes of buttons. 
   */
  variation?: ButtonVariation;
  /** 
   * Sets the HTML `type` attribute for the
   * button element.
   */ 
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"]
  /** 
   * Control the size of the button. Default is medium. 
   */
  size?: "small" | "medium" | "large";
  /** 
   * Give the button a flat texture, appropriate for forms. 
   */
  flat?: boolean;
  /** 
   * When true the button will no background; 
   * It will only have a border. 
   */
  outline?: boolean;
  /** 
   * When true, the button's width will be 100% of 
   * its container. 
   */ 
  fill?: boolean;
  /** 
   * Color the button based on the current theme. 
   */
  theme?: "light" | "dark";
  /** 
   * Make the button disabled. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates whether the button is loading some content. 
   */
  isLoading?: boolean;
  /** 
   * Sets the `className` attribute on the button element. 
   */ 
  className?: string;
  /** 
   * The function that handles click events. 
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers over the button.
   */
  onMouseEnter?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers away from the
   * button.
   */
  onMouseLeave?: React.MouseEventHandler<HTMLButtonElement>;
  /** 
   * A selector used for testing this component. 
   */
  dataTestId?: string;
  /** 
   * Content displayed inside the button. 
   */
  children?: React.ReactNode | string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({
    variation     = "primary",
    type          = "button",
    size          = "medium", 
    flat          = false,
    outline       = false,
    fill          = false, 
    theme         = "light",
    isDisabled    = false,
    isLoading     = false,
    className,
    onClick       = () => {},
    onMouseEnter  = () => {},
    onMouseLeave  = () => {},
    dataTestId,
    children}: ButtonProps,
    ref: React.Ref<HTMLButtonElement>): JSX.Element 
  {
    const childMarkup = isLoading 
      ? (
          <>
            <LoadingDots 
              dotSize={7} 
              position="center" 
              theme={loaderTheme(variation, theme, outline)} 
            />
            <span style={{opacity: '0'}}>{children}</span>
          </>
        )
      : children;

    return (
      <button
        type={type}
        className={classNames(
          className,
          styles[variation],
          styles[size],
          styles[theme],
          flat        && styles.flat,
          outline     && styles[`${variation}--outline`],
          fill        && styles.fill,
          isLoading   && styles.loading,
          isDisabled  && styles.disabled
        )}
        onClick={isDisabled ? () => {} : onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        disabled={isDisabled}
        aria-busy={isLoading}
        data-testid={dataTestId}
        ref={ref}
      >
        {childMarkup}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;