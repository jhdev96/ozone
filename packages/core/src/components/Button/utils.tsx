import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}