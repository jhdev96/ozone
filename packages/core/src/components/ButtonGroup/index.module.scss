@import '../../scss/root';


$rounded-radius: 15px;

.buttonGroup {
  display: inline-flex;
  justify-content: space-evenly;
  width: fit-content;
  box-shadow: 0px 2px 2px 2px #0000000d !important;
}

.button {
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  cursor: pointer;
  box-sizing: border-box;

  button {
    min-height: 20px;
    min-width: 20px;
    margin: 0 !important;
    padding: 10px 15px !important;
    text-shadow: none !important;
    background: none !important;
    border: 0 !important;
    border-radius: 0;
    box-shadow: none !important;

    &:focus {
      outline: 1px auto var(--gray-4) !important;
    }
  }
}

.rounded {
  border-radius: $rounded-radius;

  .button:first-child {
    border-radius: $rounded-radius 0 0 $rounded-radius;
  }

  .button:last-child {
    border-radius: 0 $rounded-radius $rounded-radius 0;
  }
}

.light {
  background-color: #fff;

  button { color: var(--black-core) !important; }

  .button {
    &:hover {
      background-color: #f2f2f2 !important;
    }
  }
}

.dark {
  background-color: var(--black-4) !important;

  button { color: var(--gray-1) !important; }

  .button {
    &:hover { 
      background-color: #2b343e !important; 
    }
  }
}