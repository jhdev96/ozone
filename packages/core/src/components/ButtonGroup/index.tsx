import React, { Children } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


export interface ButtonGroupProps {
  /** 
   * Gives the group rounded corners. 
   */
  rounded?: boolean;
  /** 
   * Gives the group a light or dark theme. 
   */
  theme?: ThemeType;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * Content to be rendered inside ButtonGroup. 
   */
  children?: React.ReactNode[];
}

export default function ButtonGroup({ 
  rounded   = true,
  theme     = 'light',
  dataTestId,
  children
}: ButtonGroupProps): JSX.Element 
{
  const childMarkup = Children.map(children, (child, index) => (
    <div className={styles.button} key={index}>
      {child}
    </div>
  ));
  
  return (
    <div 
      className={classNames(
        styles.buttonGroup,
        rounded && styles.rounded,
        styles[theme]
      )}
      data-testid={dataTestId}
    >
      {childMarkup}
    </div>
  );
}

ButtonGroup.displayName = 'ButtonGroup';