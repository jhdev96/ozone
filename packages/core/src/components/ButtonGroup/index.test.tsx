import React from 'react';
import { render, screen } from '@testing-library/react';
import ButtonGroup from '.';
import Button from '../Button';


describe('<ButtonGroup />', () => {
  it('Renders a button group with 3 buttons', () => {
    render(
      <ButtonGroup dataTestId="test">
        <Button>A</Button>
        <Button>B</Button>
        <Button>C</Button>
      </ButtonGroup>
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });

  it('renders a button group with a dark theme', () => {
    render(
      <ButtonGroup dataTestId="test" theme="dark">
        <Button>A</Button>
        <Button>B</Button>
        <Button>C</Button>
      </ButtonGroup>
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });
});