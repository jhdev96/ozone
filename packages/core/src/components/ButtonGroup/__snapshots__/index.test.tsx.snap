// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> Renders a button group with 3 buttons 1`] = `
<div
  class="buttonGroup rounded light"
  data-testid="test"
>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      A
    </button>
  </div>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      B
    </button>
  </div>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      C
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup /> renders a button group with a dark theme 1`] = `
<div
  class="buttonGroup rounded dark"
  data-testid="test"
>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      A
    </button>
  </div>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      B
    </button>
  </div>
  <div
    class="button"
  >
    <button
      aria-busy="false"
      class="primary medium light"
      type="button"
    >
      C
    </button>
  </div>
</div>
`;
